pyccel/ast/utilities.py

```diff
@@ -187,3 +194,393 @@ def split_positional_keyword_arguments(*args):
+    >>> i = Variable('int', 'i', shape=())
+    >>> d = PyccelAdd(a,b)
+    >>> expr = Assign(c,d)
+    >>> insert_index(expr, 0, i, language_has_vectors = False)
+    IndexedElement(c, i) := IndexedElement(a, i) + IndexedElement(b, i)
+    >>> insert_index(expr, 0, i, language_has_vectors = True)
+    c := a + b
+    """
+    if isinstance(expr, Variable):
+        if expr.rank==0 or -pos>expr.rank:
+            return expr
+        if expr.shape[pos]==1:
+            indexes = [Slice(None,None)]*(expr.rank+pos) + [LiteralInteger(0)]+[Slice(None,None)]*(-1-pos)
+            var = IndexedVariable(expr)
+            return var[indexes]
+    elif isinstance(expr, IndexedElement):
+        base = expr.base
+        indices = list(expr.indices)
+        if -pos>expr.base.rank or not isinstance(indices[pos], Slice):
+            return expr
+        if expr.base.shape[pos]==1:
+            assert(indices[pos].start is None)
+            indices[pos] = LiteralInteger(0)
+            return base[indices]
+        return expr
+    elif isinstance(expr, PyccelAstNode) and expr.rank==0:
+        return expr
+    elif isinstance(expr, AugAssign):
+        cls = type(expr)
+        rhs = reduce_slice_to_index(expr.rhs, pos)
+        lhs = reduce_slice_to_index(expr.lhs, pos)
+        return cls(lhs, expr.op, rhs, expr.status, expr.like)
+    elif isinstance(expr, Assign):
+        cls = type(expr)
+        rhs = reduce_slice_to_index(expr.rhs, pos)
+        lhs = reduce_slice_to_index(expr.lhs, pos)
+
+        return cls(lhs, rhs, expr.status, expr.like)
+    elif isinstance(expr, PyccelOperator):
+        cls = type(expr)
+        args = [reduce_slice_to_index(a, pos) for a in expr.args]
+        return cls(*args)
+    elif isinstance(expr, IfTernaryOperator):
+        cond        = reduce_slice_to_index(expr.cond       , pos)
+        value_true  = reduce_slice_to_index(expr.value_true , pos)
+        value_false = reduce_slice_to_index(expr.value_false, pos)
+        return IfTernaryOperator(cond, value_true, value_false)
+    return expr
+
+#==============================================================================
+def compatible_operation(*args):
+    """
+    Indicates whether an operation requires an index to be
+    correctly understood
+
+    Parameters
+    ==========
+    args : list of PyccelAstNode
+           The operator arguments
+    Results
+    =======
+    compatible : bool
+                 A boolean indicating if the operation is compatible
+    """
+    # If the shapes don't match then an index must be required
+    shapes = [a.shape for a in args if a.shape != ()]
+    shapes = set(tuple(d if isinstance(d, Literal) else -1 for d in s) for s in shapes)
+    order  = set(a.order for a in args if a.order is not None)
+    return len(shapes) == 1 and len(order) == 1
+
+#==============================================================================
+def insert_index(expr, pos, index_var, language_has_vectors):
```

Comment:
can use the function subs instead of creating a new one, this function needs to updated,  https://github.com/pyccel/pyccel/blob/master/pyccel/ast/core.py#L190, we will need to use it in multiple places, when we apply some optimizations

Reply:
I don't think we can use subs because the problem is too complicated. If you have
```
@types('int[:,:]','int[:]')
def f(a,b):
    c[:,:]=a+b
```
Then first we need to find all the variables or IndexedElements (atoms? But I think you said that doesn't work) then we need to replace them with an indexed object which preserves the Variable. The indexes aren't inserted in the same place in every variable and I need to be able to spot functions in the expression to avoid grouping objects with dependencies.

In addition I need to make sure that I don't create new objects when visiting an object (e.g. new PyccelOperator because we operate recursively on the arguments) so that I can recognise if the index was inserted or not. If it wasn't then a for loop doesn't need to be added

Reply:
atoms probably are not working anymore, it needs to be updated, but then you can use it each time you have an assignment of a vectorized expression. where you insert the for loop, all this can be done with subs unless I'm missing something.
example:
```
c = a + b
atoms = [a,b,c]
new_atoms = [a[i], b[i], c[i]]
new_expr = expr.subs(zip(atoms, new_atoms))
```

Reply:
>In addition I need to make sure that I don't create new objects when visiting an object (e.g. new PyccelOperator because we operate recursively on the arguments) so that I can recognize if the index was inserted or not. If it wasn't then a for loop doesn't need to be added.
what do you mean by that? isn't it straightforward?  you just see if it's a vectorized assignment or not, if yes you insert a for loop otherwise you don't.

Reply:
> c = a + b
atoms = [a,b,c]
new_atoms = [a[i], b[i], c[i]]
new_expr = expr.subs(zip(atoms, new_atoms))

The missing thing is mixing ranks. In my example above `a` and `c` are matrices and `b` is a vector so in fortran it would be translated to
```
c[:,i] = a[:,i] + b
```
While in c it is easier and we get
```
c[i,j]= a[i,j]+b[j]
```
So you have to work out what you're subbing and how many indices are needed

Reply:
yes, this is not an issue because you do it individually on each atom, once you get the atoms you can verify their rank and decide what to do there, and once you have your new atoms you give that to subs.
in your example, in Fortran, you will have
```
atoms = [a,b,c]
new_atoms = [a[:,i], b,  c[:,i]]
new_expr = expr.subs(zip(atoms, new_atoms))
```
and similarly with C

Reply:
if you want to make it easier for you, you can use pyccel_to_sympy, do the transformation, and then use, sympy_to_pyccel

Reply:
> yes, this is not an issue because you do it individually on each atom, once you get the atoms you can verify their rank and decide what to do there, and once you have your new atoms you give that to subs.

That makes sense but how do you handle something like
```
c += b*np.sum(a)
```
? Atoms would give us `a` but we don't want it

Reply:
Although the sum should probably be extracted from the loop anyway to stop it being calculated multiple times

Reply:
you can tell atoms where to stop, you don't search what is inside a function call for example, for the expression elimination you can check this from sympy https://github.com/sympy/sympy/blob/9e8f62e059d83178c1d8a1e19acac5473bdbf1c1/sympy/simplify/cse_main.py#L628-L772

Reply:
a lot of what you are looking for has already been done in Sympy, we either use the one of Sympy or rewrite the same functions which handle Pyccel nodes.

Reply:
> you can tell atoms where to stop

Ah I didn't realize that, in that case this might be a good answer

Reply:
So the functions we need are `atoms` and `subs`. I don't see any point in converting to sympy as we are not doing anything specific to sympy (ie calculations). So it sounds like we need to implement our own `atoms` and `subs`. I would imagine that that should be done in multiple PRs. Where would you recommend starting? Do you have any suggestions of things we can use for tests?

Reply:
we need to have an attribute  `_args` that we add to the `PyccelAstNode` that we set whenever we create a Pyccel node.
this way we can create the subs algorithm, where we go through the args, without adding all the node types.
it's also enough for the `atoms`  if a Pyccel node doesn't have args it means that it is an atom.